<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Promises</title>
</head>

<body>
    <!--
    <p>Promises are used to handle asynchronous operations in JavaScript. 
        They are easy to manage when dealing with multiple asynchronous operations where callbacks can create 
        callback hell leading to unmanageable code.</p>

    <ul>Benefits of Promises
        <li>Improves Code Readability</li>
        <li>Better handling of asynchronous operations</li>
        <li>Better flow of control definition in asynchronous logic</li>
        <li>Better Error Handling</li>
    </ul>
    
    <ol>A Promise has four states:
        <li>fulfilled: Action related to the promise succeeded</li>
        <li>rejected: Action related to the promise failed</li>
        <li>pending: Promise is still pending i.e not fulfilled or rejected yet</li>
        <li>settled: Promise has fulfilled or rejected</li>
    </ol>
    -->
    <p>A promise can be created using Promise constructor.</p>
    <!--
    Syntax :-
    var promise = new Promise(function(resolve, reject){
     //do something
    });
    -->
    <!--
        Parameters :-
    ->  Promise constructor takes only one argument,a callback function.
    ->  Callback function takes two arguments, resolve and reject
    ->  Perform operations inside the callback function and if everything went well then call resolve.
    ->  If desired operations do not go well then call reject.
    -->

    <script>
        var promise = new Promise(function (resolve, reject) {
            const x = "Coding";
            const y = "Coding";
            if (x === y) {
                resolve();
            } else {
                reject();
            }
        });

        promise.
            then(function () {
                console.log('Success, You are a Programmer');
            }).
            catch(function () {
                console.log('Some error has occured');
            }); 
    </script>

    <!--
        Promise Consumers :-
    Promises can be consumed by registering functions using .then and .catch methods.

    then()
    then() is invoked when a promise is either resolved or rejected.
    Parameters:
    then() method takes two functions as parameters.

    *  First function is executed if promise is resolved and a result is received.
    *  Second function is executed if promise is rejected and an error is received. 
    (It is optional and there is a better way to hanlde error using .catch() method


    Syntax:-
    .then(function(result){
        //handle success
    }, function(error){
        //handle error
    })

    -->

    <!--Promise Resolved-->
    <script>
        var promise = new Promise(function (resolve, reject) {
            resolve('Coding');
        })

        promise
            .then(function (successMessage) {
                //success handler function is invoked 
                console.log(successMessage);
            }, function (errorMessage) {
                console.log(errorMessage);
            }) 
    </script>

    <!--Promise Rejected-->
    <script>
        var promise = new Promise(function (resolve, reject) {
            reject('Promise Rejected')
        })

        promise
            .then(function (successMessage) {
                console.log(successMessage);
            }, function (errorMessage) {
                //error handler function is invoked 
                console.log(errorMessage);
            }) 
    </script>

    <!--
        2. catch() :-
    catch() is invoked when a promise is either rejected or some error has occured in execution.
    Parameters:
    catch() method takes one function as parameter.
    Function to handle errors or promise rejections.(.catch() method internally calls .then(null, errorHandler), i.e. .catch() is just a shorthand for .then(null, errorHandler) )
    
    Syntax:-
    .catch(function(error){
        //handle error
    })
    -->
    <p>Promise Rejected Using Catch()</p>
    <script>
        var promise = new Promise(function (resolve, reject) {
            reject('Promise Rejected')
        })

        promise
            .then(function (successMessage) {
                console.log(successMessage);
            })
            .catch(function (errorMessage) {
                //error handler function is invoked 
                console.log(errorMessage);
            }); 
    </script>

    <!--//-->

    <p>Promise Rejected Using Catch() and using throw keyword</p>
    <script>
        var promise = new Promise(function (resolve, reject) {
            throw new Error('Some error has occured')
        })

        promise
            .then(function (successMessage) {
                console.log(successMessage);
            })
            .catch(function (errorMessage) {
                //error handler function is invoked 
                console.log(errorMessage);
            }); 
    </script>

    <!--Applications-->
    <p>Applications</p>
    <!--
        Promises are used for asynchronous handling of events.
        Promises are used to handle asynchronous http requests.
    -->
</body>

</html>